org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseStartTagCodeAction.java

```diff
@@ -0,0 +1,153 @@
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.dom.LineIndentInfo;
+import org.eclipse.lemminx.services.extensions.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.utils.XMLPositionUtility;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * Code action to fix close start tag element.
+ *
+ */
+public class CloseStartTagCodeAction implements ICodeActionParticipant {
+
+	@Override
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
+			SharedSettings sharedSettings, IComponentProvider componentProvider) {
+		Range diagnosticRange = diagnostic.getRange();
+		try {
+			int startOffset = document.offsetAt(diagnosticRange.getStart());
+			DOMNode node = document.findNodeAt(startOffset);
+			if (node != null && node.isElement()) {
+				int diagnosticEndOffset = startOffset
+						+ (diagnosticRange.getEnd().getCharacter() - diagnosticRange.getStart().getCharacter());
+				DOMElement element = (DOMElement) node;
+				if (!element.hasStartTag()) {
+					// </foo>
+					DOMElement parent = element.getParentElement();
+					if (parent != null && parent.getTagName() != null) {
+						// <a><b></c>
+						// Replace "c" with "b"
+						String tagName = parent.getTagName();
+						Range replaceRange = XMLPositionUtility.selectEndTagName(element);
+						CodeAction replaceTagAction = CodeActionFactory.replace("Replace with '" + tagName + "' tag",
+								replaceRange, tagName, document.getTextDocument(), diagnostic);
+						codeActions.add(replaceTagAction);
+					}
+				} else {
+					// <foo>
+					boolean startTagClosed = element.isStartTagClosed();
+					char c = document.getText().charAt(diagnosticEndOffset - 1);
+					if (c != '/') {
+						if (startTagClosed) {
+							// ex : <foo attr="" >
+							// // Close with '</$tag>
+							String tagName = element.getTagName();
+							if (tagName != null) {
+								String label = "</" + tagName + ">";
+								String insertText = label;
+								Position endPosition = null;
+								if (!element.hasChildNodes()) {
+									int endOffset = element.getStartTagCloseOffset() + 1;
+									endPosition = document.positionAt(endOffset);
+								} else {
+									String text = document.getText();
+									// the element have some children(Text node, Element node, etc)
+									int endOffset = element.getLastChild().getEnd() - 1;
+									if (endOffset < text.length()) {
+										// remove new line
+										char ch = text.charAt(endOffset);
+										while (ch == '\r' || ch == '\n') {
+											endOffset--;
+											ch = text.charAt(endOffset);
+										}
+									}
+									endOffset++;
+									endPosition = document.positionAt(endOffset);
+									if (hasElements(element)) {

```

Comment:
I think this check should be if the last child is an element? Should we indent even if the last child is a text node?

Reply:
I have not a strong opinion about that. My idea was to manage new line for mixed content (when element to close contains an element). 

@fbricon what do you think about that?

I mean if we have

```xml
<a>
   bcd<e/>fgh
```

the code action should generate (my PR)

```xml
<a>
   bcd<e/>fgh
</a>
```

or (suggestion of @xorye )

```xml
<a>
   bcd<e/>fgh</a>
```

Reply:
indentation should match the opening tag's. So what the PR does

Reply:
ok thanks for your feedback. I let my code like this.